### PR TITLE
batch-system: use concurrent hashmap to avoid router cache

### LINKED ISSUE #15430: TiKV's memory usage keeps increases even when region count decreases and block cache cap reaches.
### What did happened?
vEz0hCBQRA
Mc09ckIhHN

### PR DESCRIPTION
**Pull request by Connor1996**
### What is changed and how it works?

Issue Number: Close #15430

What's Changed:
With investigation, we observed that the batch-system router is cloned 2000+ times at start, because each grpc method handler needs to clone a kv service instance which turns out a four times clone of router for each service instance. And each router instance has its own cache. 
5j6eWL7TtH

When one peer is destroyed, the caches in those routers can't be cleaned actively. So it's possible to occupy a lot of memory footprints. To avoid that, this PR use a lock-free hashmap to get rid of cache.

The lock-free hashmap `Dashmap` has a good enough read heavy performance https://github.com/xacrimon/conc-map-bench#read-heavy-std-hasher

```commit-message
Use concurrent hashmap to avoid router cache occupying too much memory
```

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
With this PR, memory doesn't keep growing anymore

Comparing the original implementation with cache, no considerable performance difference
image

### Release note 

```release-note
Use concurrent hashmap to avoid router cache occupying too much memory
```

### CODE REVIEW

**Comment by overvenus** [comment]
> The lock-free hashmap Dashmap

It's actually not lock-free, see https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html#method.get_mut

**Comment by Connor1996** [comment]
> > The lock-free hashmap Dashmap
> 
> It's actually not lock-free, see https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html#method.get_mut

I see, change to the word "concurrent"

**Review by tonyxuqqi — APPROVED** [review]

File: components/batch-system/src/router.rs [hunk 1]
```diff
@@ -106,72 +91,32 @@ where
...
     /// Returns None means there is no mailbox inside the normal registry.
     /// Some(None) means there is expected mailbox inside the normal registry
     /// but it returns None after apply the given function. Some(Some) means
-    /// the given function returns Some and cache is updated if it's invalid.
+    /// the given function returns Some.
     #[inline]
     fn check_do<F, R>(&self, addr: u64, mut f: F) -> CheckDoResult<R>
     where
-        F: FnMut(&BasicMailbox<N>) -> Option<R>,
+        F: FnMut(BasicMailbox<N>) -> Option<R>,
```
Comment (overvenus) [inline components/batch-system/src/router.rs:101]: Can we keep `&BasicMailbox<N>`? So that we don't need to clone mailbox for every message.

**Review by overvenus — APPROVED** [review]
Rest LGTM

File: components/batch-system/src/router.rs [hunk 2]
```diff
@@ -346,59 +279,36 @@ where
...
-    pub fn alive_cnt(&self) -> Arc<AtomicUsize> {
-        self.normals.lock().unwrap().alive_cnt.clone()
+    pub fn alive_cnt(&self) -> usize {
+        self.normals.len()
     }
 
     pub fn trace(&self) -> RouterTrace {
-        let alive = self.normals.lock().unwrap().alive_cnt.clone();
+        let alive = self.alive_cnt();
         let total = self.state_cnt.load(Ordering::Relaxed);
```
Comment (overvenus) [inline components/batch-system/src/router.rs:297]: Dead code? seems it's the only place that uses `state_cnt`.

File: components/batch-system/src/router.rs [hunk 3]
```diff
@@ -346,59 +279,36 @@ where
     /// Close the mailbox of address.
     pub fn close(&self, addr: u64) {
         info!("shutdown mailbox"; "region_id" => addr);
-        unsafe { &mut *self.caches.as_ptr() }.remove(&addr);
-        let mut mailboxes = self.normals.lock().unwrap();
-        if let Some(mb) = mailboxes.map.remove(&addr) {
+        if let Some((_, mb)) = self.normals.remove(&addr) {
```
Comment (overvenus) [inline components/batch-system/src/router.rs:282]: Maybe we need to shrink the map.

File: components/batch-system/src/router.rs (hunk 2 above)
Comment (Connor1996) [reply components/batch-system/src/router.rs:302]: in `apply.rs`, it calls `state_cnt().clone()`

**Pull request merged**